routing/fake_ending.cpp

```diff
@@ -0,0 +1,77 @@
+#include "routing/fake_ending.hpp"
+
+#include "routing/edge_estimator.hpp"
+#include "routing/index_graph.hpp"
+#include "routing/world_graph.hpp"
+
+#include "geometry/distance.hpp"
+
+#include "base/math.hpp"
+
+using namespace routing;
+using namespace std;
+
+namespace
+{
+template <typename Estimator>
+Junction CalcProjectionToSegment(Junction const & begin, Junction const & end,
+                                 m2::PointD const & point, Estimator && estimator)
+{
+  m2::ProjectionToSection<m2::PointD> projection;
+
+  projection.SetBounds(begin.GetPoint(), end.GetPoint());
+  auto const projectedPoint = projection(point);
+  auto const distBeginToEnd = estimator(begin.GetPoint(), end.GetPoint());
+
+  double constexpr kEpsMeters = 2.0;
+  if (my::AlmostEqualAbs(distBeginToEnd, 0.0, kEpsMeters))
+    return Junction(projectedPoint, begin.GetAltitude());
+
+  auto const distBeginToProjection = estimator(begin.GetPoint(), projectedPoint);
+  auto const altitude = begin.GetAltitude() + (end.GetAltitude() - begin.GetAltitude()) *
+                                                  distBeginToProjection / distBeginToEnd;
+  return Junction(projectedPoint, altitude);
+}
+
+template <typename Estimator>
+FakeEnding MakeFakeEndingImpl(Junction const & backJunction, Junction const & frontJunction,
+                              Segment const & segment, m2::PointD const & point, bool oneWay,
+                              Estimator && estimator)
+{
+  auto const & projectedJunction =
+      CalcProjectionToSegment(backJunction, frontJunction, point, estimator);
```

Comment:
If you use estimator only here, please, use perfect forwarding.

Reply:
done